ws-client/src/transport.rs

```diff
@@ -184,74 +187,135 @@ impl<'a> WsTransportClientBuilder<'a> {
+					Ok(stream) => stream,
+					Err(e) => {
+						// TODO(niklasad1): this should be debug.
+						log::error!("Failed to connect to sockaddr: {:?}", sockaddr);
+						err = Some(Err(e));
+						continue;
 					}
+				};
+				let mut client = WsHandshakeClient::new(
+					BufReader::new(BufWriter::new(tcp_stream)),
+					&target.host_header,
+					&target.path_and_query,
+				);
+				if let Some(origin) = &origin {
+					client.set_headers(origin);
 				}
-				future::Either::Right((_, _)) => return Err(WsHandshakeError::Timeout(self.timeout)),
-			}
-		};
+				// Perform the initial handshake.
+				match client.handshake().await {
+					Ok(ServerResponse::Accepted { .. }) => {
+						// TODO(niklasad1): this should be debug.
+						log::error!("Connection established to target: {:?}", target);
+						let mut builder = client.into_builder();
+						builder.set_max_message_size(self.max_request_body_size as usize);
+						let (sender, receiver) = builder.finish();
+						return Ok((Sender { inner: sender }, Receiver { inner: receiver }));
+					}
 
-		log::debug!("Connecting to target: {:?}", self.target);
-		let mut client = WsRawClient::new(
-			BufReader::new(BufWriter::new(tcp_stream)),
-			&self.target.host_header,
-			&self.target.path_and_query,
-		);
-		if let Some(origin) = self.origin_header.as_ref() {
-			client.set_origin(origin);
+					Ok(ServerResponse::Rejected { status_code }) => {
+						// TODO(niklasad1): this should be debug.
+						log::error!("Connection rejected: {:?}", status_code);
+						err = Some(Err(WsHandshakeError::Rejected { status_code }));
+					}
+					Ok(ServerResponse::Redirect { status_code, location }) => {
+						log::error!("Redirection: status_code: {}, location: {}", status_code, location);
+						match url::Url::parse(&location) {
+							// redirection with absolute path => need to lookup.
+							Ok(url) => {
+								target = Target::parse(url)?;
+								tls_connector = match target.mode {
+									Mode::Tls => {
+										let mut client_config = rustls::ClientConfig::default();
+										if let CertificateStore::Native = self.certificate_store {
+											client_config.root_store = rustls_native_certs::load_native_certs()
+												.map_err(|(_, e)| WsHandshakeError::CertificateStore(e))?;
+										}
+										Some(Arc::new(client_config).into())
+									}
+									Mode::Plain => None,
+								};
+								break;
+							}
+							// redirection is relative, either `/baz` or `bar`.
+							Err(
+								url::ParseError::RelativeUrlWithoutBase
+								| url::ParseError::RelativeUrlWithCannotBeABaseBase,
+							) => {
+								// replace the entire path if `location` is `/`.
+								if location.starts_with('/') {
+									target.path_and_query = location;
+								} else {
+									// join paths such that the leaf is replaced with `location`.
+									let strip_last_child = Path::new(&target.path_and_query)
```

Comment:
we need to have a slightly modified version of this because on Windows the paths are appended with `\\` instead of `/` which is not according to the HTTP relative path.